File: repeated-substring-pattern/Solution459.java
// https://leetcode.com/problems/repeated-substring-pattern/

class Solution459 {

    // T: O(n^2), S: O(n)
    public boolean repeatedSubstringPattern(String s) {
        int n = s.length();
        for (int i = 1; i <= n / 2; i++) {
            if (n % i == 0) {
                var substring = s.substring(0, i);
                var sb = new StringBuilder();
                for (int j = 0; j < n / i; j++) {
                    sb.append(substring);
                }
                if (sb.toString().equals(s)) {
                    return true;
                }
            }
        }
        return false;
    }

    // T: O(n), S: O(n)
    public boolean repeatedSubstringPatternOpt(String s) {
        var t = s + s;
        var sub = t.substring(1, t.length() - 1);
        return sub.contains(s);
    }

    public static void main(String[] args) {
        var sol = new Solution459();
        System.out.println(sol.repeatedSubstringPattern("abab")); // true
        System.out.println(sol.repeatedSubstringPattern("aba")); // false
        System.out.println(sol.repeatedSubstringPattern("abcabcabcabc")); // true
        System.out.println(sol.repeatedSubstringPatternOpt("abcabcabcabc")); // true
    }

}

File: repeated-substring-pattern/solution459.py
# https://leetcode.com/problems/repeated-substring-pattern/

class Solution459:

    # T: O(n^2), S: O(n)
    def repeatedSubstringPattern(self, s: str) -> bool:
        n = len(s)
        for i in range(1, n // 2 + 1):
            if n % i == 0:
                sub = s[:i]
                sb = sub * (n // i)
                if sb == s:
                    return True
        return False

    # T: O(n), S: O(n)
    def repeatedSubstringPatternOpt(self, s: str) -> bool:
        t = s + s
        sub = t[1:-1]
        return s in sub


def main() -> None:
    sol = Solution459()
    print(sol.repeatedSubstringPattern("abab"))  # True
    print(sol.repeatedSubstringPattern("aba"))  # False
    print(sol.repeatedSubstringPattern("abcabcabcabc"))  # True
    print(sol.repeatedSubstringPatternOpt("abcabcabcabc"))  # True


if __name__ == "__main__":
    main()
